Added beta specs docs (#1977)
holochain/holochain: docs/specs/src/hwp_3_correctness.md

to refer to external docs on DeepKey and DPKI\]

Evolvability
------------

For large scale systems to work well over time, we contend that specific
architectural elements and affordances make a significant difference in
their capacity to evolve while maintaining overall coherence as they do
so:

1.  **Subsidiarity:** From the Wikipedia definition: "*Subsidiarity is a
  principle of social organization that holds that social and
  political issues should be dealt with at the most immediate (or
  local) level that is consistent with their resolution.*"
  Subsidiarity enhances evolvability because it insulates the whole
  system from too much change, while simultaneously allowing change
  where it is needed. Architecturally, however, subsidiarity is not
  easy to implement because it is rarely immediately obvious what
  level of any system is consistent with an issue's resolution.\
  \
  In Holochain, the principle of subsidiarity is embodied many ways,
  but crucially in the architecture of app instances having fully
  separate DNAs running on their own separate networks, each also
  having clear and differentiable Integrity and Coordination

>   specifications. This creates a very clear loci of change, both at

Reviewer: ```suggestion
  specifications. This creates very clear loci of change, both at
```

System Correctness: Confidence
==============================

In the frame of the Byzantine Generals problem, the correctness of a
distributed coordination system is analyzed through the lens of "fault
tolerance". In our frame we take on a broader scope and address the

> question of the many kinds of confidence necessary for a system adoption

Reviewer: ```suggestion
question of the many kinds of confidence necessary for a system's adoption
```

> and continued use. We identify and address the following dimensions
> necessary confidence:

Reviewer: ```suggestion
and continued use. We identify and address the following dimensions
of confidence:
```

1.  **Fault Tolerance:** the system's resilience to external
  perturbations, both malicious and natural. Intrinsic integrity.

> 2.  **Completeness/Fit:** the system's apriori design elements that

Reviewer: ```suggestion
2.  **Completeness/Fit:** the system's *a priori* design elements that
```

  demonstrate fitness for purpose. We demonstrate this by describing
  how Holochain addresses, multi-agent reality binding, scalability,
  and shared-state finality.

3.  **Security:** the system's resilience to intentional disruption by

>   malicious action.

Reviewer: point 1, Fault Tolerance, already covers malicious disruption, so how is this different?

Author: It feels to me like there's a qualitative difference between gaming behavior and ability to gate access to parts of the systems through things like our capabilities model, that is of a different order of intentional distruption than the kind of fault-tolerance that happens from cosmic rays that flips bits (which you could also do intentionally).  The only reason I put "intentional" under fault tolerance is that one can't necessarily know if a bit-flip disruption event was intentional or random, but one can pretty easily know that a cosimic ray didn't orchestrate a 51% attack.

This is meant to make the distinction between those two types.  Any idea how better to describe that?

  malicious or incautious actors from stealing or revealing private
  keys, however, the system does include affordances to deal with
  these realities. These are discussed under Completeness/Fit.

2.  **Faults from data corruptibility in transmission and storage:**
  Because all state data is stored along with a cryptographic hash
  of that data, and because all data is addressed and retrieved by
  that hash and can be compared against the retrieved data, the only
  possible fault is that the corruption resulted in data that has
  the same hash. For Sha256 hashing (which is what we use), this is
  known to be a vanishingly small possibility for both intentional
  and unintentional data corruption.[^6] Furthermore because all
  data is stored as hash-chains, it is not possible for portions of
  data to be retroactively changed. Agent's Source Chains thus
  become immutable append-only event logs.\
  \
  One possible malicious act that an Agent can take is to roll back
  their chain to some point and start publishing different data from
  that point forward. But, because the publishing protocol requires
  Agents to publish Actions to the neighborhood of their own public

>   key, any actions that lead to a forked chain will be easily and
>   immediately detected by simple matching more than one action
>   linked to a previous action.\

Reviewer: ```suggestion
  key, any actions that lead to a forked chain will be easily and
  immediately detected by simply detecting more than one action
  linked to the same previous action.\
```

b.  **Cryptocurrencies Based on P2P Accounting:** Extending the previous
  example, if Alice wants to transfer 100 units of a currency to
  Bob, they can both sign a single entry where Alice is in the
  spender role, and Bob the receiver. This provides similar
  guarantees as familiar double-entry accounting, ensuring changes
  happen to both accounts simultaneously. You can easily compute
  someone's balance by a replay of the transactions on their source
  chain, and you can hold both signing parties accountable for any
  fraudulent transfers breaking the data integrity rules of the
  currency application. There's no need for global time of
  transactions when each is clearly ordered by its sequence in the
  chains of the only accounts affected by the change.

c.  **Witnessed Authoritative Sequence:** Some applications may require
  an authoritative sequence of changes to a specific data type.
  Consider changes to membership of a group of administrators, where
  Carol and David are both members of the group, and Carol commits a
  change which removes David, and David commits a change which
  removes Carol from the group. With no global time clock to trust,

>   who's change wins? An application can set up a small pool of

Reviewer: ```suggestion
  whose change wins? An application can set up a small pool of
```

  witnesses and configure their countersigning session to require M

>   50% of N) and whichever action the

Reviewer: This "M 50% of N" needs to be cleaned up

  to tentatively solidify a state of affairs when they see gossip
  for a time window has calmed and neighbors have converged on the
  same state. After this settling period (which might get set to
  something between 5 to 15 minutes) any later changes which would
  produce a conflict (such as forking a chain) can be rejected.

Security
--------

The system's resilience to intentional gaming, and disruption by
malicious actors will be covered in depth in future papers, but here we
provide an overview.

There are many contributing factors to what allows systems to live up to
the varying safety and security requirements of their users. In general,
the approach taken in Holochain is to provide affordances that take into
account the many types of real-world costs that result from adding
security and safety to systems such that application developers can
match the trade-offs of those costs to their application context. The
integrity guarantees listed in the formal system description, detail the
fundamental data safety that Holochain applications provide. Some other
important facets of system security and safety come from:

> 1.  Gating access functions that change state, for which Holochain

Reviewer: ```suggestion
1.  Gating access to functions that change state, for which Holochain
```

  provides a unified and flexible Object Capabilities model

2.  Detecting and Blocking participation of bad actors, for which
  Holochain provides the affordances of validation and warranting.

3.  Protection from Attack categories

4.  Resilience to Human Error

### Gating Access via Cryptographic Object Capabilities

> To use a Holochain application end-users must trigger zome calls that

Reviewer: This is the first time "zome" has been introduced. The term should be introduced (capitalized) somewhere earlier, probably when introducing DNA. By the way, "DNAs" (as a countable noun) was not really introduced as such.

later lookup. The \"agent\" type grant is just the agent\'s public key.

### Validation & Warranting

We have already covered how Holochain's agent-centric validation and
intrinsic data integrity provides security from malicious actors trying
to introduce invalid or incorrect information into an Application's
network as every agent can deterministically verify data and thus
self-secure. It is also important, however, to be able to eject
malicious actors who generate or propagate such invalid data from
network participation so as to secure against the resource drain that
such action could incur.

More details on warranting \[Todo: ACB\] \[expand on: one good apple
heals the bunch\]

### Security from Attack Categories

#### Consensus​ ​Attacks

This whole category of attack starts from the assumption that consensus
is required for distributed systems. Because Holochain doesn't start
from that assumption the attack category really doesn't apply, but it's
worth mentioning because there​ ​are​ ​a​ ​number​ ​of​ ​attacks​ ​on​
​blockchain​ ​which​ ​disrupt their distributed computing solution by
having collusion between the majority of the nodes. ​The​ ​usual

> thinking​ ​is​ ​that​ ​it​ ​takes​ ​a​ ​large​ ​number​ ​of​ ​nodes45​

Reviewer: ```suggestion
thinking​ ​is​ ​that​ ​it​ ​takes​ ​a​ ​large​ ​number​ ​of​ ​nodes
```

​and​ ​massive​ ​amounts​ ​of​ ​computing​ ​power or financial

> incentives46​ ​to prevent​ ​undue​ ​hijacking​ ​of​ ​consensus.​

Reviewer: ```suggestion
incentives​ ​to prevent​ ​undue​ ​hijacking​ ​of​ ​consensus.​
```

Reviewer: I see some other random numbers scattered around, please check for them in general

​However,​ ​since​ ​Holochain's data coherence doesn't derive from all
nodes proceeding in consensus​ lockstep,​ ​but​ rather ​on​ determinist
validation, nobody​ ​ever​ ​needs​ ​to​ ​trust​ ​a​ ​consensus​
​lottery.​ ​

#### Sybil Attacks

Membranes & DPKI \[TODO: ACB\] \[expand on: membrane proofs\] \[expand
on: Holochain enables​ ​continuity​ ​of​ ​identity​ ​across​
​application​ ​contexts​ ​with​ ​its​ ​DPKI​ ​app,44​ ​which​ ​can​
​interface with​ ​decentralized​ ​identity​ ​services​ ​of​ ​your​
​choosing.\]

#### Denial-of-Service Attacks

Holochain is not systemically subject to denial-of-service attacks
because there is no central point to attack. Because each application is
its own network, attackers would have to flood every agent of every
application to carry out a systemic denial-of-service attack and to do
what would require knowing who all those agents are which is also not

> recorded anywhere. One point of vulnerability are the boot-strap servers

Reviewer: bootstrap can be one word everywhere

for an application. But this is not a systemic vulnerability as each
application can designate its own boot-strap server, and they can also
be arbitrarily hardened against denial-of-service to suit the needs of
the application.

#### Spamming Attacks

> Holochain includes a native rate-limiting on entry creation \[TODO:
> ACB\]

Reviewer: ```suggestion
Holochain includes a native mechanism for rate-limiting an agent's actions,
which is enforced by the same integrity layer which verifies cryptographic
and structural integrity of actions.

Using the [generic cell rate algorithm](https://en.wikipedia.org/wiki/Generic_cell_rate_algorithm),
each agent is allotted a certain amount of capacity which is replenished over time.
Each action taken has a certain "weight" assigned to it which depletes 
available capacity accordingly. Any action which would exceed available capacity is
deemed invalid. The available capacity of an agent at any given moment can be 
deterministically and independently verified by any agent on the network
who can view the agent's entire chain of actions, so per-agent rate limits
can be strictly enforced through the validation and warranting mechanism.
```
